### PR TITLE
Updater uses informers to fetch pod controllers.

### PR DESCRIPTION
**Pull request by bskiba**
This includes SatefulSets, Replicasets and ReplicationContollers.

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go [hunk 1]
```diff
@@ -220,35 +263,54 @@ func getPodID(pod *apiv1.Pod) string {
...
-func getReplicaCount(creator podReplicaCreator, client kube_client.Interface) (int, error) {
+func (f *podsEvictionRestrictionFactoryImpl) getReplicaCount(creator podReplicaCreator) (int, error) {
 	switch creator.Kind {
-	case "ReplicationController":
-		rc, err := client.CoreV1().ReplicationControllers(creator.Namespace).Get(creator.Name, metav1.GetOptions{})
-
-		if err != nil || rc == nil {
+	case replicationController:
+		rcObj, exists, err := f.rcInformer.GetStore().GetByKey(creator.Namespace + "/" + creator.Name)
+		if err != nil || rcObj == nil {
```
Comment (kgolab) [inline vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go]: Shouldn't the == nil check be done together with !exists.

File: vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go [hunk 2]
```diff
@@ -220,35 +263,54 @@ func getPodID(pod *apiv1.Pod) string {
...
 		}
 		return int(*rc.Spec.Replicas), nil
 
-	case "ReplicaSet":
-		rs, err := client.ExtensionsV1beta1().ReplicaSets(creator.Namespace).Get(creator.Name, metav1.GetOptions{})
-
-		if err != nil || rs == nil {
+	case replicaSet:
+		rsObj, exists, err := f.rsInformer.GetStore().GetByKey(creator.Namespace + "/" + creator.Name)
+		if err != nil || rsObj == nil {
```
Comment (kgolab) [inline vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go]: Same as for line 270.

File: vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go [hunk 3]
```diff
@@ -113,9 +134,9 @@ func (e *podsEvictionRestrictionImpl) Evict(podToEvict *apiv1.Pod, eventRecorder
 			Name:      podToEvict.Name,
 		},
 	}
-	err := e.client.CoreV1().Pods(podToEvict.Namespace).Evict(eviction)
+	err := e.client.CoreV1().Pods(eviction.Namespace).Evict(eviction)
```
Comment (kgolab) [inline vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go]: This seems unrelated (and does not change the code behaviour).

Comment (bskiba) [reply vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go]: Done.

File: vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go (hunk 1 above)
Comment (bskiba) [reply vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go]: I think this check is redundant anyway. Removed.

**Comment by bskiba** [comment]
Also fixed failing test

File: vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go [hunk 4]
```diff
@@ -61,14 +70,26 @@ type PodsEvictionRestrictionFactory interface {
 
 type podsEvictionRestrictionFactoryImpl struct {
 	client                    kube_client.Interface
+	rcInformer                cache.SharedIndexInformer
```
Comment (kgolab) [inline vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go]: Could you please either extend the names here or at least comment them? Looking below it's easy to understand what they mean but without this context it's not clear so let's have some context directly here.

Comment (bskiba) [reply vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go]: Done
